README.md
[<img src="./images/prysma.png" height="180">](https://github.com/Rooknj/prysma)

# Prysma-UI

[![Build Status](https://travis-ci.org/Rooknj/prysma-ui.svg?branch=master)](https://travis-ci.org/Rooknj/prysma-ui)
[![Renovate enabled](https://img.shields.io/badge/renovate-enabled-brightgreen.svg)](https://renovatebot.com/)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

User Interface for Prysma

## Installing

Prysma-UI is bundled with Prysma and is served statically using express.

Currently new releases of Prysma-UI are manually added to Prysma.

## Developing

### Built With

- [React](https://github.com/facebook/react)
- [Apollo](https://www.apollographql.com/docs/react/)
- [React-Router](https://reacttraining.com/react-router/)
- [Styled-Components](https://www.styled-components.com/)
- [Material-UI](https://material-ui.com/)

### Prerequisites

- Nodejs v8.10 or greater is required as this project uses [Create React App](https://facebook.github.io/create-react-app/docs/getting-started#creating-an-app) (Check by running `node -v`)
- Yarn (Check by running `yarn -v`)

### Setting up Dev

```
git clone https://github.com/Rooknj/prysma-ui.git
cd prysma-ui
yarn install
```

### Run Prysma-UI locally

Using API specified through environment configuration (defaults to http://prysma.local:80)

Default Config Environment Variables

- PROXY_PROTOCOL=http
- PROXY_HOST=prysma.local
- PROXY_PORT=80

```
yarn start
```

Using API located at http://localhost:4001 (the default location for a locally running instance of prysma)

```
yarn start --local
```

### Syncing GraphQL Schema

We use [GraphQL Code Generator](https://graphql-code-generator.com/) to keep the client and server beautifully in sync

To sync, make sure you have a Prysma server running at http://localhost:4001/graphql then run `yarn codegen`

Configuration is located at [codegen.yml](/codegen.yml)

### Testing

Tests are written using [Jest](https://jestjs.io/)

Run tests on watch mode:

```
yarn test
```

Run tests with coverage:

```
yarn test --coverage
```

Run tests once:

```
yarn test --no-watch
```

### Building

Create Production Bundle (outputs to /build)

```
yarn build
```

### Deploying

Deployments happen automatically using Semantic-Release based off of semantic commit messages

## Commit Messages

Commit messages must follow [Conventional Commits](https://www.conventionalcommits.org)

Example Messages
- chore(docs): updated Readme
- refactor: renamed LedStrip to Light
- fix(LightMessenger): fixed a bug where an incorrect message format was being used
- feat: Added ability to assign lights to different rooms
- BREAKING CHANGE: Swapped database from Redis to Sqlite3

CI will fail if you do not use the correct commit format. Local messages are linted using [Commitlint](https://commitlint.js.org/#/)

Commitizen support coming soon

## Versioning

We use [SemVer 2.0.0](https://semver.org/) for versioning. To see available versions, check out the [releases page](https://github.com/Rooknj/prysma-ui/releases)

## Style Guide

### Linting

Linting is managed by [ESLint](https://eslint.org/)

Run `yarn lint` to check all files

Currently we are extending eslint:recommended, @typescript-eslint/recommended, airbnb, eslint-comments/recommended, and jest/recommended

Full config is located in [.eslintrc.js](/.eslintrc.js)

Note: The linter is ran during CI and the build will fail if there are any errors. Currently we are not automatically fixing linting errors so you will have to make sure the linter passes manually

### Code Style

Code style is managed by [Prettier](https://prettier.io/) [![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)

All default prettier styles are enabled except for the overrides located in [.prettierrc.js](/.prettierrc.js)

Prettier is automatically ran on staged files every commit using [Husky](https://github.com/typicode/husky) and [Lint-Staged](https://github.com/okonet/lint-staged)

You can run prettier manually on every file through the command `yarn reformat` or you can set up prettier to run inside your IDE/Editor

## License

This project is licensed under the terms of the
[MIT license](/LICENSE).


CHANGELOG.md
## [1.0.30](https://github.com/Rooknj/prysma-ui/compare/v1.0.29...v1.0.30) (2019-10-03)


### Bug Fixes

* **deps:** update react monorepo to v16.10.2 ([8a6f6ad](https://github.com/Rooknj/prysma-ui/commit/8a6f6ad))

## [1.0.29](https://github.com/Rooknj/prysma-ui/compare/v1.0.28...v1.0.29) (2019-10-02)


### Bug Fixes

* **deps:** update apollo graphql packages to v3.1.2 ([de2424e](https://github.com/Rooknj/prysma-ui/commit/de2424e))

## [1.0.28](https://github.com/Rooknj/prysma-ui/compare/v1.0.27...v1.0.28) (2019-10-01)


### Bug Fixes

* **deps:** update dependency react-router-dom to v5.1.2 ([6995786](https://github.com/Rooknj/prysma-ui/commit/6995786))

## [1.0.27](https://github.com/Rooknj/prysma-ui/compare/v1.0.26...v1.0.27) (2019-09-28)


### Bug Fixes

* **deps:** update react monorepo to v16.10.1 ([e873064](https://github.com/Rooknj/prysma-ui/commit/e873064))

## [1.0.26](https://github.com/Rooknj/prysma-ui/compare/v1.0.25...v1.0.26) (2019-09-28)


### Bug Fixes

* **deps:** update dependency react-router-dom to v5.1.1 ([9409b02](https://github.com/Rooknj/prysma-ui/commit/9409b02))

## [1.0.25](https://github.com/Rooknj/prysma-ui/compare/v1.0.24...v1.0.25) (2019-09-27)


### Bug Fixes

* **deps:** update react monorepo to v16.10.0 ([f7b799b](https://github.com/Rooknj/prysma-ui/commit/f7b799b))

## [1.0.24](https://github.com/Rooknj/prysma-ui/compare/v1.0.23...v1.0.24) (2019-09-25)


### Bug Fixes

* **deps:** update dependency graphql to v14.5.8 ([64ea1ea](https://github.com/Rooknj/prysma-ui/commit/64ea1ea))

## [1.0.23](https://github.com/Rooknj/prysma-ui/compare/v1.0.22...v1.0.23) (2019-09-24)


### Bug Fixes

* **deps:** update dependency react-router-dom to v5.1.0 ([1749f61](https://github.com/Rooknj/prysma-ui/commit/1749f61))

## [1.0.22](https://github.com/Rooknj/prysma-ui/compare/v1.0.21...v1.0.22) (2019-09-23)


### Bug Fixes

* **deps:** update dependency styled-components to v4.4.0 ([4acc15f](https://github.com/Rooknj/prysma-ui/commit/4acc15f))

## [1.0.21](https://github.com/Rooknj/prysma-ui/compare/v1.0.20...v1.0.21) (2019-09-20)


### Bug Fixes

* **deps:** update dependency graphql to v14.5.7 ([c891eea](https://github.com/Rooknj/prysma-ui/commit/c891eea))

## [1.0.20](https://github.com/Rooknj/prysma-ui/compare/v1.0.19...v1.0.20) (2019-09-16)


### Bug Fixes

* **deps:** update apollo graphql packages to v3.1.1 ([305aa55](https://github.com/Rooknj/prysma-ui/commit/305aa55))

## [1.0.19](https://github.com/Rooknj/prysma-ui/compare/v1.0.18...v1.0.19) (2019-09-15)


### Bug Fixes

* **deps:** update dependency graphql to v14.5.6 ([62c2ed9](https://github.com/Rooknj/prysma-ui/commit/62c2ed9))

## [1.0.18](https://github.com/Rooknj/prysma-ui/compare/v1.0.17...v1.0.18) (2019-09-13)


### Bug Fixes

* **deps:** update dependency graphql to v14.5.5 ([a922a5e](https://github.com/Rooknj/prysma-ui/commit/a922a5e))

## [1.0.17](https://github.com/Rooknj/prysma-ui/compare/v1.0.16...v1.0.17) (2019-09-11)


### Bug Fixes

* **deps:** update dependency @material-ui/core to v4.4.2 ([fd5f3cb](https://github.com/Rooknj/prysma-ui/commit/fd5f3cb))

## [1.0.16](https://github.com/Rooknj/prysma-ui/compare/v1.0.15...v1.0.16) (2019-09-08)


### Bug Fixes

* **deps:** update mui monorepo to v4.4.1 ([ed8d467](https://github.com/Rooknj/prysma-ui/commit/ed8d467))

## [1.0.15](https://github.com/Rooknj/prysma-ui/compare/v1.0.14...v1.0.15) (2019-09-06)


### Bug Fixes

* **deps:** update apollo graphql packages ([899f9b4](https://github.com/Rooknj/prysma-ui/commit/899f9b4))

## [1.0.14](https://github.com/Rooknj/prysma-ui/compare/v1.0.13...v1.0.14) (2019-09-06)


### Bug Fixes

* **deps:** update apollo graphql packages to v3.1.0 ([f095b2b](https://github.com/Rooknj/prysma-ui/commit/f095b2b))

## [1.0.13](https://github.com/Rooknj/prysma-ui/compare/v1.0.12...v1.0.13) (2019-08-31)


### Bug Fixes

* **deps:** update dependency @material-ui/core to v4.4.0 ([25e896f](https://github.com/Rooknj/prysma-ui/commit/25e896f))

## [1.0.12](https://github.com/Rooknj/prysma-ui/compare/v1.0.11...v1.0.12) (2019-08-29)


### Bug Fixes

* **deps:** update dependency graphql to v14.5.4 ([a1f2315](https://github.com/Rooknj/prysma-ui/commit/a1f2315))

## [1.0.11](https://github.com/Rooknj/prysma-ui/compare/v1.0.10...v1.0.11) (2019-08-28)


### Bug Fixes

* **deps:** update apollo graphql packages to v3.0.1 ([2a998ad](https://github.com/Rooknj/prysma-ui/commit/2a998ad))
* **deps:** update dependency graphql to v14.5.3 ([01ff889](https://github.com/Rooknj/prysma-ui/commit/01ff889))

## [1.0.10](https://github.com/Rooknj/prysma-ui/compare/v1.0.9...v1.0.10) (2019-08-22)


### Bug Fixes

* **deps:** update mui monorepo to v4.3.3 ([bf57092](https://github.com/Rooknj/prysma-ui/commit/bf57092))

## [1.0.9](https://github.com/Rooknj/prysma-ui/compare/v1.0.8...v1.0.9) (2019-08-11)


### Bug Fixes

* **deps:** update apollo graphql packages ([b3c7cac](https://github.com/Rooknj/prysma-ui/commit/b3c7cac))
* **deps:** update dependency @material-ui/core to v4.3.2 ([2df2319](https://github.com/Rooknj/prysma-ui/commit/2df2319))
* **deps:** update dependency typeface-roboto to v0.0.75 ([774ad83](https://github.com/Rooknj/prysma-ui/commit/774ad83))
* **deps:** update react monorepo to v16.9.0 ([cd42370](https://github.com/Rooknj/prysma-ui/commit/cd42370))

## [1.0.8](https://github.com/Rooknj/prysma-ui/compare/v1.0.7...v1.0.8) (2019-08-08)


### Bug Fixes

* **deps:** update dependency @material-ui/core to v4.3.1 ([78b5bfa](https://github.com/Rooknj/prysma-ui/commit/78b5bfa))

## [1.0.7](https://github.com/Rooknj/prysma-ui/compare/v1.0.6...v1.0.7) (2019-08-07)


### Bug Fixes

* **deps:** upgraded to react-apollo 3 ([9b2d020](https://github.com/Rooknj/prysma-ui/commit/9b2d020))

## [1.0.6](https://github.com/Rooknj/prysma-ui/compare/v1.0.5...v1.0.6) (2019-08-01)


### Bug Fixes

* **deps:** update mui monorepo to v4.3.0 ([2e5d57b](https://github.com/Rooknj/prysma-ui/commit/2e5d57b))


scripts/start.ts
/* eslint no-console:0 */
/* eslint import/no-extraneous-dependencies: ["error", {"devDependencies": true}] */
import { argv } from "yargs";
import execa from "execa";

process.env.REACT_APP_ENV = "development";

const { PROXY_PROTOCOL = "http", PROXY_HOST = "prysma.local", PROXY_PORT = "80" } = process.env;

// Crash on unhandled rejections
process.on("unhandledRejection", (err): never => {
  throw err;
});

const startArgs = process.argv.slice(2);

if (argv.local) {
  // Remove --local from argv
  startArgs.splice(startArgs.indexOf("--local"), 1);

  // Set the env variable to use the local server
  console.log("Using Server at http://localhost:4001");
  process.env.PROXY = "http://localhost:4001";
} else {
  process.env.PROXY = `${PROXY_PROTOCOL}://${PROXY_HOST}:${PROXY_PORT}`;
  console.log(`Using Server at ${process.env.PROXY}`);
}

execa.sync("react-scripts", ["start", ...startArgs], {
  stdio: "inherit",
  preferLocal: true,
});
